File: ruby/lib/svix/models/stream_sink_patch.rb
# frozen_string_literal: true
# This file is @generated
require "json"

require_relative "./azure_blob_storage_config_patch"
require_relative "./big_query_config_patch"
require_relative "./clickhouse_config_patch"
require_relative "./event_bridge_config_patch"
require_relative "./google_cloud_pub_sub_config_patch"
require_relative "./google_cloud_storage_config_patch"
require_relative "./otel_tracing_config_patch"
require_relative "./rabbit_mq_config_patch"
require_relative "./redshift_config_patch"
require_relative "./s3_config_patch"
require_relative "./sink_http_config_patch"
require_relative "./sink_status_in"
require_relative "./snowflake_config_patch"
require_relative "./sns_config_patch"
require_relative "./sqs_config_patch"

module Svix
  class StreamSinkPatchConfig
    class Poller

      def serialize
        Hash.new
      end

      def self.deserialize(attributes = {})
        new
      end
      # Serializes the object to a json string
      # @return String
      def to_json(*args)
        serialize.to_json(*args)
      end
    end

    class AzureBlobStorage < AzureBlobStorageConfigPatch
    end

    class OtelTracing < OtelTracingConfigPatch
    end

    class Http < SinkHttpConfigPatch
    end

    class AmazonS3 < S3ConfigPatch
    end

    class GoogleCloudStorage < GoogleCloudStorageConfigPatch
    end

    class GoogleCloudPubSub < GoogleCloudPubSubConfigPatch
    end

    class Sqs < SqsConfigPatch
    end

    class Sns < SnsConfigPatch
    end

    class BigQuery < BigQueryConfigPatch
    end

    class Clickhouse < ClickhouseConfigPatch
    end

    class EventBridge < EventBridgeConfigPatch
    end

    class Snowflake < SnowflakeConfigPatch
    end

    class RabbitMq < RabbitMqConfigPatch
    end

    class Redshift < RedshiftConfigPatch
    end
  end

  class StreamSinkPatch
    # The StreamSink's UID.
    attr_accessor :uid
    attr_accessor :status
    attr_accessor :batch_size
    attr_accessor :max_wait_secs
    attr_accessor :event_types
    attr_accessor :channels
    attr_accessor :metadata
    attr_accessor :config

    ALL_FIELD ||= ["uid", "status", "batch_size", "max_wait_secs", "event_types", "channels", "metadata", "config"].freeze
    private_constant :ALL_FIELD
    TYPE_TO_NAME = {
      StreamSinkPatchConfig::Poller => "poller",
      StreamSinkPatchConfig::AzureBlobStorage => "azureBlobStorage",
      StreamSinkPatchConfig::OtelTracing => "otelTracing",
      StreamSinkPatchConfig::Http => "http",
      StreamSinkPatchConfig::AmazonS3 => "amazonS3",
      StreamSinkPatchConfig::GoogleCloudStorage => "googleCloudStorage",
      StreamSinkPatchConfig::GoogleCloudPubSub => "googleCloudPubSub",
      StreamSinkPatchConfig::Sqs => "sqs",
      StreamSinkPatchConfig::Sns => "sns",
      StreamSinkPatchConfig::BigQuery => "bigQuery",
      StreamSinkPatchConfig::Clickhouse => "clickhouse",
      StreamSinkPatchConfig::EventBridge => "eventBridge",
      StreamSinkPatchConfig::Snowflake => "snowflake",
      StreamSinkPatchConfig::RabbitMq => "rabbitMq",
      StreamSinkPatchConfig::Redshift => "redshift"
    }
    private_constant :TYPE_TO_NAME
    NAME_TO_TYPE = TYPE_TO_NAME.invert
    private_constant :NAME_TO_TYPE

    def initialize(attributes = {})
      unless attributes.is_a?(Hash)
        fail(
          ArgumentError,
          "The input argument (attributes) must be a hash in `Svix::StreamSinkPatch` new method"
        )
      end

      attributes.each do |k, v|
        unless ALL_FIELD.include?(k.to_s)
          fail(ArgumentError, "The field #{k} is not part of Svix::StreamSinkPatch")
        end

        if k.to_s == "config"
          unless TYPE_TO_NAME.key?(v.class)
            fail(ArgumentError, "The field #{k} can't be a `#{v.class}` expected one of #{TYPE_TO_NAME.keys}")
          end

          instance_variable_set("@__enum_discriminator", TYPE_TO_NAME[v.class])
        end

        instance_variable_set("@#{k}", v)
        instance_variable_set("@__#{k}_is_defined", true)
      end

      if @__enum_discriminator.nil?
        fail(ArgumentError, "Required config field was not set")
      end
    end

    def self.deserialize(attributes = {})
      attributes = attributes.transform_keys(&:to_s)
      attrs = Hash.new
      attrs["uid"] = attributes["uid"]
      attrs["status"] = Svix::SinkStatusIn.deserialize(attributes["status"]) if attributes["status"]
      attrs["batch_size"] = attributes["batchSize"]
      attrs["max_wait_secs"] = attributes["maxWaitSecs"]
      attrs["event_types"] = attributes["eventTypes"]
      attrs["channels"] = attributes["channels"]
      attrs["metadata"] = attributes["metadata"]
      unless NAME_TO_TYPE.key?(attributes["type"])
        fail(ArgumentError, "Invalid type `#{attributes["type"]}` expected on of #{NAME_TO_TYPE.keys}")
      end

      unless attributes.key?("config")
        fail(ArgumentError, "Missing required field config")
      end

      attrs["config"] = NAME_TO_TYPE[attributes["type"]].deserialize(attributes["config"])
      new(attrs)
    end

    def serialize
      out = Hash.new
      out["uid"] = Svix::serialize_primitive(@uid) if @__uid_is_defined
      out["status"] = Svix::serialize_schema_ref(@status) if @__status_is_defined
      out["batchSize"] = Svix::serialize_primitive(@batch_size) if @__batch_size_is_defined
      out["maxWaitSecs"] = Svix::serialize_primitive(@max_wait_secs) if @__max_wait_secs_is_defined
      out["eventTypes"] = Svix::serialize_primitive(@event_types) if @event_types
      out["channels"] = Svix::serialize_primitive(@channels) if @channels
      out["metadata"] = Svix::serialize_primitive(@metadata) if @metadata
      out["type"] = @__enum_discriminator
      out["config"] = @config.serialize
      out
    end

    # Serializes the object to a json string
    # @return String
    def to_json(*args)
      serialize.to_json(*args)
    end

  end
end

File: python/svix/models/stream_sink_patch.py
# this file is @generated
import typing as t

from pydantic import ModelWrapValidatorHandler, model_validator
from typing_extensions import Self

from .azure_blob_storage_config_patch import AzureBlobStorageConfigPatch
from .big_query_config_patch import BigQueryConfigPatch
from .clickhouse_config_patch import ClickhouseConfigPatch
from .common import BaseModel
from .event_bridge_config_patch import EventBridgeConfigPatch
from .google_cloud_pub_sub_config_patch import GoogleCloudPubSubConfigPatch
from .google_cloud_storage_config_patch import GoogleCloudStorageConfigPatch
from .otel_tracing_config_patch import OtelTracingConfigPatch
from .rabbit_mq_config_patch import RabbitMqConfigPatch
from .redshift_config_patch import RedshiftConfigPatch
from .s3_config_patch import S3ConfigPatch
from .sink_http_config_patch import SinkHttpConfigPatch
from .sink_status_in import SinkStatusIn
from .snowflake_config_patch import SnowflakeConfigPatch
from .sns_config_patch import SnsConfigPatch
from .sqs_config_patch import SqsConfigPatch


class StreamSinkPatch(BaseModel):
    uid: t.Optional[str] = None
    """The StreamSink's UID."""

    status: t.Optional[SinkStatusIn] = None

    batch_size: t.Optional[int] = None

    max_wait_secs: t.Optional[int] = None

    event_types: t.Optional[t.List[str]] = None

    channels: t.Optional[t.List[str]] = None

    metadata: t.Optional[t.Dict[str, str]] = None

    type: t.Union[
        t.Literal["poller"],
        t.Literal["azureBlobStorage"],
        t.Literal["otelTracing"],
        t.Literal["http"],
        t.Literal["amazonS3"],
        t.Literal["googleCloudStorage"],
        t.Literal["googleCloudPubSub"],
        t.Literal["sqs"],
        t.Literal["sns"],
        t.Literal["bigQuery"],
        t.Literal["clickhouse"],
        t.Literal["eventBridge"],
        t.Literal["snowflake"],
        t.Literal["rabbitMq"],
        t.Literal["redshift"],
    ]
    config: t.Union[
        t.Dict[str, t.Any],
        AzureBlobStorageConfigPatch,
        OtelTracingConfigPatch,
        SinkHttpConfigPatch,
        S3ConfigPatch,
        GoogleCloudStorageConfigPatch,
        GoogleCloudPubSubConfigPatch,
        SqsConfigPatch,
        SnsConfigPatch,
        BigQueryConfigPatch,
        ClickhouseConfigPatch,
        EventBridgeConfigPatch,
        SnowflakeConfigPatch,
        RabbitMqConfigPatch,
        RedshiftConfigPatch,
    ]

    @model_validator(mode="wrap")
    @classmethod
    def validate_model(
        cls, data: t.Any, handler: ModelWrapValidatorHandler[Self]
    ) -> Self:
        if isinstance(data, cls):
            return handler(data)
        if "config" not in data:
            data["config"] = {}
        output = handler(data)
        if output.type == "poller":
            output.config = data.get("config", {})
        elif output.type == "azureBlobStorage":
            output.config = AzureBlobStorageConfigPatch.model_validate(
                data.get("config", {})
            )
        elif output.type == "otelTracing":
            output.config = OtelTracingConfigPatch.model_validate(
                data.get("config", {})
            )
        elif output.type == "http":
            output.config = SinkHttpConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "amazonS3":
            output.config = S3ConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "googleCloudStorage":
            output.config = GoogleCloudStorageConfigPatch.model_validate(
                data.get("config", {})
            )
        elif output.type == "googleCloudPubSub":
            output.config = GoogleCloudPubSubConfigPatch.model_validate(
                data.get("config", {})
            )
        elif output.type == "sqs":
            output.config = SqsConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "sns":
            output.config = SnsConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "bigQuery":
            output.config = BigQueryConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "clickhouse":
            output.config = ClickhouseConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "eventBridge":
            output.config = EventBridgeConfigPatch.model_validate(
                data.get("config", {})
            )
        elif output.type == "snowflake":
            output.config = SnowflakeConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "rabbitMq":
            output.config = RabbitMqConfigPatch.model_validate(data.get("config", {}))
        elif output.type == "redshift":
            output.config = RedshiftConfigPatch.model_validate(data.get("config", {}))
        else:
            raise ValueError(f"Unexpected type `{output.type}`")
        return output
